### DISCUSSION TITLE
Immer set typescript issue

### DISCUSSION THREAD

**Discussion by kaisiant** [pmndrs/zustand#2346] [Q&A] [answered]
Hello guys, I need some help on typescript for immer.

My code: 
```
import { create } from "zustand";
import { persist, createJSONStorage } from "zustand/middleware";
import { immer } from "zustand/middleware/immer";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Region } from "@medusajs/medusa";

interface AppState {
  region?: Region;
}

interface AppAction {
  setRegion: (region: Region) => void;
}

export const useAppStore = create<AppState & AppAction>()(
  persist(
    immer((set) => ({
      region: undefined,
      setRegion: (region: Region) =>
        set((state) => {
          state.region = region;
        }),
    })),
    { name: "appStore", storage: createJSONStorage(() => AsyncStorage) }
  )
);

```

<img width="649" alt="image" src="https://github.com/pmndrs/zustand/assets/26655111/95f2a00d-5a6c-4a64-906a-954a3778b366">

Not able to figure out which part I setup incorrectly.
Referred to this https://tsplay.dev/wja7kW and other resources, mostly are outdated.
And is it possible to use persist, immer and combine? If yes, how to set them up, combine wrap immer or immer wrap combine.
Thanks.

**Comment by dbritto-dev** [comment]
@kaisiant does this help? -> https://docs.pmnd.rs/zustand/guides/typescript

**Reply by kaisiant** [reply]
I tried this https://docs.pmnd.rs/zustand/integrations/immer-middleware, the set is not typed as well
```
import { create } from 'zustand'
import { immer } from 'zustand/middleware/immer'

type State = {
  count: number
}

type Actions = {
  increment: (qty: number) => void
  decrement: (qty: number) => void
}

export const useCountStore = create<State & Actions>()(
  immer((set) => ({
    count: 0,
    increment: (qty: number) =>
      set((state) => {
        state.count += qty
      }),
    decrement: (qty: number) =>
      set((state) => {
        state.count -= qty
      }),
  })),
)
```

**Reply by dbritto-dev** [reply]
I don't see any errors -> https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgHIAvAVwGcYMA7AEzIChRJZE5QQ0o8hYqUq16zAPQhgTJgBs0AdwzoJvfm1YwAnmBwBletjgBeRKzgoIVBjABccBlRAAjfq1ystunAEFkMMAQDDSm5pbADKhofLYOABQAjjoOTq78AJSmAHxwAG4QMhZwTGjRsfZwSSmOzm5QWSa5BUWerGgAHpzwyMF0cLRoAMLWtobQOGbRWGgAPIYzcABkcP6BfdnxGfHF6lDx8TRoMI258QjFlr02lQAMADSXPFHoFQnJ2ql1mTlPlkcwA50GancKWcH-IxoAB011scAA1GYPn88BlHhDSuU0HEqh8vukGr8If9jkCoaCLiTwcDsLDRvAALTInSo3Do4rsjkZVhAA

**Reply by kaisiant** [reply]
May I know which version you are using?
React Native CSB:https://codesandbox.io/p/sandbox/peaceful-shamir-vl62fv
React CSB: https://codesandbox.io/p/sandbox/hardcore-pine-y7h2ks

**Reply by kaisiant** [reply, marked as answer]
Too bad, I just realized forgot to run `npm install immer`.
